### PR TITLE
Interceptor for identity resolution when attaching graphs

### LINKED ISSUE #20124: Provide a hook for identity resolution
In general, identity resolution is a difficult problem. See https://github.com/dotnet/efcore/issues/20116#issuecomment-593499871 and many other similar comments over the years.

However, maybe we could provide a hook that applications can do to implement their own identity resolution when EF starts tracking a graph. This likely wouldn't cover difficult were merging needs to happen, but simple cases where duplicates can be essentially ignored could become much easier.

### PR DESCRIPTION
**Pull request by ajcvickers**
Part of #626
Fixes #20124

This is essentially a call back that happens when an instance with the same key as an existing tracked instance is attached. The existing instance is always retained, but the interceptor gets the chance to change property values of the tracked instance. This will typically be used to copy over values from the duplicate instance (i.e. painting from a detached entity) or just skip the duplicate. Interceptor implementations are provided to do these two things.

### CODE REVIEW

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs [hunk 1]
```diff
@@ -428,6 +428,17 @@ private void SetServiceProperties(EntityState oldState, EntityState newState)
         }
     }
 
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public bool ResolveToExistingEntry(
```
Comment (AndriySvyryd) [inline src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs:437]: Why is it needed here? Calling it on StateManager doesn't seem too bad

File: src/EFCore/Diagnostics/SkippingIdentityResolutionInterceptor.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.EntityFrameworkCore.Diagnostics;
+
+/// <summary>
+///     A <see cref="IIdentityResolutionInterceptor"/> that ignores the new instance and retains property values from the existing
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/SkippingIdentityResolutionInterceptor.cs:7]: Isn't this just the default behavior?

File: src/EFCore/Diagnostics/CopyingIdentityResolutionInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.EntityFrameworkCore.Diagnostics;
+
+/// <summary>
+///     A <see cref="IIdentityResolutionInterceptor"/> that copies property values from the new entity instance into the
+///     tracked entity instance.
+/// </summary>
+public class CopyingIdentityResolutionInterceptor : IIdentityResolutionInterceptor
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/CopyingIdentityResolutionInterceptor.cs:10]: OverridingIdentityResolutionInterceptor?

Comment (AndriySvyryd) [reply src/EFCore/Diagnostics/CopyingIdentityResolutionInterceptor.cs:10]: Perhaps we could add an option to not override modified values

Comment (AndriySvyryd) [reply src/EFCore/Diagnostics/CopyingIdentityResolutionInterceptor.cs:10]: And another option to also copy over original values

File: test/EFCore.Specification.Tests/GraphUpdates/GraphUpdatesTestBaseMiscellaneous.cs [hunk 4]
```diff
@@ -1570,4 +1569,132 @@ public virtual void Can_add_multiple_dependents_when_multiple_possible_principal
                         + hiddenAreaTask.Choices.Count(e => e.Id == taskChoice.Id));
                 }
             });
+
+    [ConditionalFact]
+    public void Cab_attach_full_required_graph_of_duplicates()
```
Comment (AndriySvyryd) [inline test/EFCore.Specification.Tests/GraphUpdates/GraphUpdatesTestBaseMiscellaneous.cs:1574]:
```suggestion
    public void Can_attach_full_required_graph_of_duplicates()
```
Also in the other tests

File: test/EFCore.Specification.Tests/GraphUpdates/GraphUpdatesTestBaseMiscellaneous.cs [hunk 5]
```diff
@@ -1570,4 +1569,132 @@ public virtual void Can_add_multiple_dependents_when_multiple_possible_principal
...
+                var trackedRoot = LoadRequiredCompositeGraph(context);
+                var entries = context.ChangeTracker.Entries().ToList();
+
+                context.Attach(QueryRequiredCompositeGraph(context).AsNoTracking().Single(IsTheRoot));
+
+                AssertEntries(entries, context.ChangeTracker.Entries().ToList());
+                AssertNavigations(trackedRoot);
+
+                Assert.Equal(0, context.SaveChanges());
+            });
```
Comment (AndriySvyryd) [inline test/EFCore.Specification.Tests/GraphUpdates/GraphUpdatesTestBaseMiscellaneous.cs:1699]: Also add tests for skip navs

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore/Diagnostics/CopyingIdentityResolutionInterceptor.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore/Diagnostics/CopyingIdentityResolutionInterceptor.cs:10]: Done.

File: src/EFCore/Diagnostics/SkippingIdentityResolutionInterceptor.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore/Diagnostics/SkippingIdentityResolutionInterceptor.cs:7]: I've been thinking about that a lot--whether it is exactly the same thing as this. Today I came to the conclusion that it was different enough to not just be this. Primarily, I think it is something that should be done per query, not set as an interceptor on the context. We can discuss in a design meeting.

File: test/EFCore.Specification.Tests/GraphUpdates/GraphUpdatesTestBaseMiscellaneous.cs (hunk 5 above)
Comment (ajcvickers) [reply test/EFCore.Specification.Tests/GraphUpdates/GraphUpdatesTestBaseMiscellaneous.cs:1699]: Done.

**Pull request merged**